File: src/main/java/helper/MultiLanguageDecision.java
package helper;

import learners.LanguageDecision;
import main.Learning;

import java.util.HashMap;
import java.util.Map;

/***
 * Represents a language decision where there's multiple languages or multiple weighted
 * parts in each decision.
 */
public class MultiLanguageDecision extends LanguageDecision {

    private final Map<String, Double> items = new HashMap<>();


    /** Adds weight to the specified language. */
    private void addWeightTo(String item, double weight) {
        // insert into the map if not already there
        if (!items.containsKey(item))
            items.put(item, 0.0);

        // add on the weight
        items.put(item, items.get(item) + weight);
    }

    /**
     * Adds the weight to the overall language decision, with this decision
     * being just as useful as the other ones.
     * @param decide The language decision for the part.
     */
    public void addWeightTo(LanguageDecision decide) {
        addWeightTo(decide, 1.0);
    }

    /**
     * Adds the weight to the overall language decision, with this decision playing
     * just a fraction of the overall decision.
     * @param decide The language decision for the part.
     * @param weight The relative weight of this part's decision.
     */
    public void addWeightTo(LanguageDecision decide, double weight) {
        for (String language : Learning.languages) {
            addWeightTo(language, decide.confidenceForLanguage(language) * weight);
        }
    }

    /**
     * Normalizes the weights so that the sum is 1.0
     */
    public void normalize() {
        double totalWeight = items.values().stream().mapToDouble(i -> i).sum();
        for (String key : items.keySet()) {
            items.put(key, items.get(key) / totalWeight);
        }
    }

    @Override
    public double confidenceForLanguage(String language) {
        if (items.containsKey(language))
            return items.get(language);
        return 0;
    }

    @Override
    public String mostConfidentLanguage() {
        String mostCommon = Learning.languages.get(0);
        double maxConfidence = 0;
        for (String i : Learning.languages) {
            double c = confidenceForLanguage(i);
            if (c > maxConfidence) {
                mostCommon = i;
                maxConfidence = c;
            }
        }
        return mostCommon;
    }
}

File: src/main/kotlin/com/zachjones/languageclassifier/entities/MultiLanguageDecision.kt
package com.zachjones.languageclassifier.entities

import com.zachjones.languageclassifier.model.types.Language

/***
 * Represents a language decision where there's multiple languages or multiple weighted
 * parts in each decision.
 */
data class MultiLanguageDecision(
    val weights: Map<Language, Double>
) : LanguageDecision() {

    /**
     * Normalizes the weights so that the sum is 1.0
     */
    fun normalized(): MultiLanguageDecision {
        val totalWeight = weights.values.sum()
        return MultiLanguageDecision(weights = weights.mapValues { (_, value) ->
            value / totalWeight
        })
    }

    override fun confidences(): Map<Language, Double> = weights

    companion object {
        fun of(decisions: List<LanguageDecision>): MultiLanguageDecision {
            val map = mutableMapOf<Language, Double>()
            Language.values().forEach {
                map[it] = 0.0
            }
            decisions.forEach { decision ->
                decision.confidences().map { (language, confidence) ->
                    map[language] = map[language]!! + confidence
                }
            }
            // remove all the "votes" for other, and let the majority language win
            map[Language.OTHER] = 0.0
            return MultiLanguageDecision(map).normalized()
        }
    }
}

